contracts/DXswapFeeReceiver.sol

```diff
@@ -6,122 +6,226 @@ import './interfaces/IWETH.sol';
 import './libraries/TransferHelper.sol';
 import './libraries/SafeMath.sol';
 
-
 contract DXswapFeeReceiver {
-    using SafeMath for uint;
+    using SafeMath for uint256;
 
     address public owner;
     IDXswapFactory public factory;
     address public WETH;
     address public ethReceiver;
     address public fallbackReceiver;
+    uint32 public maxSwapPriceImpact = 100; // uses default 1% as max allowed price impact for takeProtocolFee swap
+
+    // if needed set address of external project which can get % of total earned protocol fee
+    // % of total protocol fee to external project (100 means 1%) is within the range <0, 50>
+    struct ExternalFeeReceiver {
+        address externalReceiver;
+        uint32 feePercentage;
+    }
+
+    mapping(address => ExternalFeeReceiver) public externalFeeReceivers;
+
+    event TakeProtocolFee(address indexed sender, address indexed to, uint256 NumberOfPairs);
 
     constructor(
-        address _owner, address _factory, address _WETH, address _ethReceiver, address _fallbackReceiver
+        address _owner,
+        address _factory,
+        address _WETH,
+        address _ethReceiver,
+        address _fallbackReceiver
     ) public {
         owner = _owner;
         factory = IDXswapFactory(_factory);
         WETH = _WETH;
         ethReceiver = _ethReceiver;
         fallbackReceiver = _fallbackReceiver;
     }
-    
+
     function() external payable {}
 
+    // called by the owner to set the new owner
     function transferOwnership(address newOwner) external {
         require(msg.sender == owner, 'DXswapFeeReceiver: FORBIDDEN');
         owner = newOwner;
     }
-    
+
+    // called by the owner to change receivers addresses
     function changeReceivers(address _ethReceiver, address _fallbackReceiver) external {
         require(msg.sender == owner, 'DXswapFeeReceiver: FORBIDDEN');
         ethReceiver = _ethReceiver;
         fallbackReceiver = _fallbackReceiver;
     }
-    
+
     // Returns sorted token addresses, used to handle return values from pairs sorted in this order
     function sortTokens(address tokenA, address tokenB) internal pure returns (address token0, address token1) {
         require(tokenA != tokenB, 'DXswapFeeReceiver: IDENTICAL_ADDRESSES');
         (token0, token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
         require(token0 != address(0), 'DXswapFeeReceiver: ZERO_ADDRESS');
     }
-    
+
     // Helper function to know if an address is a contract, extcodesize returns the size of the code of a smart
     //  contract in a specific address
-    function isContract(address addr) internal returns (bool) {
-        uint size;
-        assembly { size := extcodesize(addr) }
+    function isContract(address addr) internal view returns (bool) {
```

Comment:
This must be replaced by a call to the factory. Increases gas consumption a little bit, but since it's only used to check if a pair exists, in the process we also have guarantees about the validity of a pair (and we avoid surprises like the hack we've had).